src/ReachSets/ContinuousPost/BFFPS19/reach_blocks.jl

```diff
@@ -0,0 +1,281 @@
+
+    X_store = (output_function == nothing) ?
+                array :
+                box_approximation(output_function(array))
+
+    return X_store
+end
+
+# sparse
+function reach_blocks!(ϕ::SparseMatrixCSC{NUM, Int},
+                       Xhat0::Vector{<:LazySet{NUM}},
+                       U::Union{ConstantInput, Nothing},
+                       overapproximate::Function,
+                       overapproximate_inputs::Function,
+                       n::Int,
+                       N::Union{Int, Nothing},
+                       output_function::Union{Function, Nothing},
+                       blocks::AbstractVector{Int},
+                       partition::AbstractVector{<:Union{AbstractVector{Int}, Int}},
+                       δ::NUM,
+                       guards_proj::Vector{<:LazySet{NUM}},
+                       block_options,
+                       vars::Vector{Int},
+                       termination::Function,
+                       progress_meter::Union{Progress, Nothing},
+                       res::Vector{<:ReachSet}
+                       )::Tuple{Int, Bool} where {NUM}
+    update!(progress_meter, 1)
+    array = CartesianProductArray(Xhat0[blocks])
+    X_store = (output_function == nothing) ?
+        array :
+        box_approximation(output_function(array))
+    t0 = zero(δ)
+    t1 = δ
+
+    diff_blocks = setdiff(1:length(partition),blocks)
+
+    b = length(blocks)
+    bd = length(diff_blocks)
+
+    array_d = CartesianProductArray(Xhat0[diff_blocks])
+    X_store_d = (output_function == nothing) ?
+            array_d :
+            box_approximation(output_function(array_d))
+
+    terminate, skip, X_store = termination_X(1, X_store, t0, X_store_d, blocks,
+                                diff_blocks, block_options, termination)
+    store!(res, 1, X_store, t0, t1, N)
+    if terminate
+        return 1, skip
+    end
+
+    b = length(blocks)
+    Xhatk = Vector{LazySet{NUM}}(undef, b)
+    Xhatk_d = Vector{LazySet{NUM}}(undef, bd)
+    ϕpowerk = copy(ϕ)
+
+    if U != nothing
+        Whatk_blocks = Vector{LazySet{NUM}}(undef, b)
+        Whatk_diff_blocks = Vector{LazySet{NUM}}(undef, bd)
+        inputs = next_set(U)
+        @inbounds for i in 1:b
+            bi = partition[blocks[i]]
+            Whatk_blocks[i] = overapproximate_inputs(1, blocks[i], proj(bi, n) * inputs)
+        end
+        @inbounds for i in 1:bd
+            bi = partition[diff_blocks[i]]
+            Whatk_diff_blocks[i] = overapproximate_inputs(1, diff_blocks[i], proj(bi, n) * inputs)
+        end
+    end
+
+    k = 2
+    @inbounds while true
+        update!(progress_meter, k)
+        X_store = deco_post_sparse(b, blocks, Whatk_blocks, partition,
+                                          ϕpowerk, Xhatk, Xhat0, output_function, overapproximate)
+
+        t0 = t1
+        t1 += δ
+
+        terminate, skip, reach_set_intersected = termination(k, X_store, t0)
+        if reach_set_intersected isa EmptySet
```

Comment:
```suggestion
        if isempty(reach_set_intersected)
           break
        end
```
or even `isempty(reach_set_intersected) && break` (same in the other places).

Reply:
Okay since we always operate on bounded sets and the emptiness information is given in the type. Generally `isempty` is more costly.

Reply:
The termination makes an `isempty` check (see eg. `termination_inv_out`), and a new  `isempty` check is made here. This is intended? Why not using the first output argument (`terminate`) here?

Reply:
> The termination makes an isempty check

Not always. You can also terminate by bounding the number of steps.

Reply:
Hmm and you can't use the second argument, `skip`? What is the purpose of [these checks](https://github.com/JuliaReach/Reachability.jl/pull/641/files#diff-0cf932a40c31c44ed8396aab06ec361eR279)? Adding documentation would help the maintainability of this code :)